lib/ddtrace/configuration/validation_proxy.rb

```diff
@@ -0,0 +1,101 @@
+require 'forwardable'
+
+module Datadog
+  module Configuration
+    # Forwards configuration settings that are permitted,
+    # but raises errors for access to anything else.
+    class ValidationProxy
+      extend Forwardable
+
+      FORWARDED_METHODS = [
+        :reset!,
+        :to_h
+      ].freeze
+
+      def_delegators \
+        :settings,
+        *FORWARDED_METHODS
+
+      def initialize(settings)
+        @settings = settings
+      end
+
+      protected
+
+      attr_reader :settings
+
+      # Forwards global configuration settings
+      class Global < self
+        FORWARDED_METHODS = [
+          :api_key,
+          :api_key=,
+          :diagnostics,
+          :env,
+          :env=,
+          :logger,
+          :service,
+          :service_without_fallback,
+          :service=,
+          :site,
+          :site=,
+          :tags,
+          :tags=,
+          :time_now_provider,
+          :time_now_provider=,
+          :version,
+          :version=
+        ].freeze
+
+        def_delegators \
+          :settings,
+          *FORWARDED_METHODS
+      end
```

Comment:
It looks like very subclass of `ValidationProxy` is subclassing only to provide its own `FORWARDED_METHODS`. Which leads me to think this could be a parameter of the class itself:

```ruby
[1] pry(main)> value = "123"
=> "123"
[7] pry(main)> class ValidationProxy
[7] pry(main)*   extend Forwardable
[7] pry(main)*   def initialize(target:, methods:)
[7] pry(main)*     @target = target
[7] pry(main)*     singleton_class.def_delegator(:@target, *methods)
[7] pry(main)*   end
[7] pry(main)* end
=> :initialize
[8] pry(main)> p1 = ValidationProxy.new(target: value, methods: [:to_s])
=> #<ValidationProxy:0x00007fcc61a17888 @target="123">
[9] pry(main)> p1.to_s
=> "123"
[10] pry(main)> p2 = ValidationProxy.new(target: value, methods: [:to_i])
=> #<ValidationProxy:0x00007fcc52258a48 @target="123">
[11] pry(main)> p2.to_s
=> "#<ValidationProxy:0x00007fcc52258a48>"
[12] pry(main)> p2.to_i
=> 123
```

then, we could have

```ruby
CI_METHODS = [:ci_mode]
validator = ValidationProxy.new(target: something, methods: CI_METHODS)
```

Reply:
I was thinking about doing exactly something like this, but didn't have the time to find the right articulation to make this work. I may shamelessly steal and implement a form of this.

Reply:
Do keep in mind that (I think) use of `singleton_class` does impart a performance hit. So doing this each time we instantiate a proxy (which is every time we access configuration) could be expensive.

Reply:
Good point, I wasn't considering that there's a few places that do a lot of settings lookups 👍